csrc/scheduler/matmul.cpp

```diff
@@ -51,6 +57,364 @@ void moveInnerBroadcastLeft(TensorView* tv, int number_of_inner_pos = 3) {
+    // TODO: right now, we are assuming ldmatrix access, which only supports
+    // sizeof(T) == 16bit (i.e. half/bfloat16) load according to offical doc:
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#warp-level-matrix-load-instruction-ldmatrix
+    // In the future, when we start adding support for tf32(different macro),
+    // fp32(ffma), double, int8, fp8, etc. we need to update this function.
+    TORCH_INTERNAL_ASSERT(dataTypeSize(*shared_mem_tv->getDataType()) == 2);
+
+    // ldmatrix loads a ldmatrix_rows x ldmatrix_cols = 8 x 8 matrix each time,
+    constexpr int ldmatrix_rows = 8;
+    constexpr int ldmatrix_cols = 8;
+
+    // Column size of the tile needs to be multiples of 8 for ldmatrix to work.
+    TORCH_INTERNAL_ASSERT(
+        tile_size_x >= ldmatrix_rows && tile_size_x % ldmatrix_rows == 0 &&
+            tile_size_y >= ldmatrix_cols && tile_size_y % ldmatrix_cols == 0,
+        "Prolog swizzle for ldmatrix, illegal tile size for prolog swizzle",
+        tile_size_x,
+        "x",
+        tile_size_y);
+
+    /* Note [How to remove bank conflict for ldmatrix?]
+     *
+     * **This note is interleaved with code, I suggest reading this note like
+     *   reading a jupyter notebook**
+     *
+     * Our task is to make sure different rows does not fall into the same
+     * bank of shared memory.
+     *
+     * Introduction to bank conflict can be found at page 54-72 of:
+     * https://on-demand.gputechconf.com/gtc/2018/presentation/s81006-volta-architecture-and-performance-optimization.pdf
+     *
+     * When we talk about bank conflict removal, we are talking about the
+     * following task:
+     *   "there are 32 banks, and each bank contains one 4-byte word, we want to
+     *    make sure different lanes in a warp does not access different word
+     *    addresses in the same bank"
+     * For example, if thread 0 is accessing word address 1, and thread 1 is
+     * accessing word address 33, then these two threads will have a bank
+     * conflict because they are accessing different word addresses in the same
+     * bank. However, if thread 0 is accessing byte address 4 and thread 1 is
+     * accessing byte address 6 then there will be no bank conflict because 4
+     * and 6 both belong to word 1.
+     */
+
+    constexpr int smem_bytes_per_word = 4;
+    constexpr int smem_banks = 32;
+
+    /* but here, for our convenience, because ldmatrix always use vectorized
+     * access of 8 items = 16 bytes = 4 words, we further group words into
+     * units: we consider each 4 words as a "unit", and each 4 banks as a
+     * "megabank". So we can rephrase our task as:
+     *   "there are 8 megabanks, and each megabanks contains one 4-word unit, we
+     *    want to make sure different lanes in a warp does not access different
+     *    unit addresses in the same megabank"
+     * In this terminology, each matrix has 8 rows, and each row has exactly one
```

Comment:
Thanks for the detailed writeup. Maybe it's also good to mention the matrix is in the row major format.